### ISSUE TITLE
Exception when updating `data.List.foldRight` in @unison/base/main

### ISSUE THREAD

**Issue by jskripsky** [unisonweb/unison#4362]
I'm using UCM version eb83404. 

# Exception
```
Encountered exception:
[malformed type] IntroOuter not in lexical scope of matching Forall
CallStack (from HasCallStack):
  error, called at src/Unison/KindInference/Generate.hs:145:18 in unison-parser-typechecker-0.0.0-6MGtlMk5aXuJWsQZ16HFt2:Unison.KindInference.Generate
```

# Unison Code
```Haskell
data.List.foldRight : (a ->{𝕖} b ->{𝕖} b) -> b -> [a] ->{𝕖} b
data.List.foldRight f b =
  go b = cases
    as :+ a -> go (f a b) as
    [] -> b
  go b
```

# How to reproduce
- `clone @unison/base`
- paste the code above into scratch file
- `update`

Adding the same function with a different name works, only updating causes this exception.

**Comment by ceedubs** [comment]
This looks related to #4354. At least, the error message looks similar. That this one is reported to only happen on update is surprising to me.

**Comment by jskripsky** [comment]
Downgrading UCM to release/M5g "fixes" the problem, just like in #4354.
